src/TransformationUtils.cc

```diff
@@ -88,57 +88,63 @@ std::unique_ptr<TransitionSystem> toTransitionSystem(ChcDirectedGraph const & gr
+void strongConnection(std::unordered_set<SymRef, SymRefHash> & visitedVertices,
+                      std::unordered_set<SymRef, SymRefHash> & verticesOnStack,
                       AdjacencyListsGraphRepresentation const & graphRepresentation, ChcDirectedGraph const & graph,
-                      SymRef node) {
-    visitedVertices.insert(node.x);
-    verticesOnStack.insert(node.x);
+                      SymRef node, std::vector<EId> & loop) {
+    visitedVertices.insert(node);
+    verticesOnStack.insert(node);
     auto const & outEdges = graphRepresentation.getOutgoingEdgesFor(node);
     if (size(outEdges) <= 1) {
-        verticesOnStack.erase(node.x);
-        return false;
+        verticesOnStack.erase(node);
+        return;
     }
 
     for (EId eid : outEdges) {
         if (graph.getTarget(eid) != node) {
             auto nextVertex = graph.getTarget(eid);
-            if (visitedVertices.find(nextVertex.x) == visitedVertices.end()) {
-                bool loopFound =
-                    strongConnection(visitedVertices, verticesOnStack, graphRepresentation, graph, nextVertex);
-                if (loopFound) { return true; }
-            } else if (verticesOnStack.find(nextVertex.x) != verticesOnStack.end()) {
-                return true;
+            if (visitedVertices.find(nextVertex) == visitedVertices.end()) {
+                strongConnection(visitedVertices, verticesOnStack, graphRepresentation, graph, nextVertex, loop);
+                if (!loop.empty() && graph.getTarget(eid) != graph.getTarget(loop[0]) &&
+                    graph.getTarget(eid) == graph.getSource(loop[loop.size() - 1])) {
+                    loop.push_back(eid);
+                    return;
+                }
+            } else if (verticesOnStack.find(nextVertex) != verticesOnStack.end()) {
+                loop.push_back(eid);
+                return;
             }
         }
     }
 
-    verticesOnStack.erase(node.x);
-
-    return false;
+    verticesOnStack.erase(node);
 }
 
-bool TarjanLoopDetection(ChcDirectedGraph const & graph) {
+std::vector<EId> detectLoop(const ChcDirectedGraph & graph) {
+    std::vector<EId> loop;
+    if (graph.getVertices().size() <= 3) { return loop; }
     auto graphRepresentation = AdjacencyListsGraphRepresentation::from(graph);
     auto vertices = reversePostOrder(graph, graphRepresentation);
-    std::unordered_set<int> visitedVertices;
-    std::unordered_set<int> verticesOnStack;
-
-    for (uint i = 1; i < vertices.size() - 1; i++) {
-        if (visitedVertices.find(vertices[i].x) == visitedVertices.end()) {
-            bool loop_detected =
-                strongConnection(visitedVertices, verticesOnStack, graphRepresentation, graph, vertices[i]);
-            if (loop_detected) { return true; }
+    std::unordered_set<SymRef, SymRefHash> visitedVertices;
+    std::unordered_set<SymRef, SymRefHash> verticesOnStack;
+    for (uint i = 1; i < vertices.size() - 1 && loop.size() == 0; i++) {
```

Comment:
Can you move the condition `loop.size() == 0` to the body of the loop. Here it is somewhat hidden.

Actually, wouldn't it be sufficient to run `strongConnection` on the entry of the graph?

Reply:
Updated